storage/innobase/handler/ha_innodb.cc

```diff
@@ -13282,6 +13288,86 @@ ha_innobase::rename_table(
 
 	DBUG_RETURN(convert_error_code_to_mysql(error, 0, NULL));
 }
+/********************************************************************//**
+Initialize committed count within dict_table_t
+@return 0 or error code
+*/
+
+int
+ha_innobase::enable_persistent_count()
+{
+	dict_table_t* ib_table = m_prebuilt->table;
+	uchar* buf = m_prebuilt->m_mysql_table->record[0];
+	trx_t* trx = m_prebuilt->trx;
+	int err;
+
+	ib_table->alter_persistent_count = true;
+	mutex_enter(&ib_table->committed_count_mutex);
+
+	if (ib_table->committed_count_inited) {
+		mutex_exit(&ib_table->committed_count_mutex);
+		return -1;  /* Already initialized */
+	}
+
+	ib_table->committed_count = 0;
+	rnd_init(true);
+	do {
+		err = rnd_next(buf);
+		if (!err) {
+			ib_table->committed_count++;
+		}
+	} while (!err);
+	ib_table->committed_count -= trx->uncommitted_count(ib_table);
+	ib_table->committed_count_inited = true;
+	m_int_table_flags |= HA_PERSISTENT_COUNT;
+	cached_table_flags = table_flags();
+
+	mutex_exit(&ib_table->committed_count_mutex);
```

Comment:
Here, we seem to be holding a mutex while scanning an entire table. Mutexes should only be held for a short time, preferrably not at all during I/O operations.

Reply:
Holding `dict_index_t:lock` could also be unacceptable. Infamously, during `CHECK TABLE` InnoDB is increasing the default hang detection watchdog timeout from 600 seconds, because `btr_validate_level()` is holding the index latch. I think that we should rely on MDL here. And we might want to create a custom function for counting the records more quickly, without invoking `rec_get_offsets()` at all.